src/main/commandServer/__tests__/settingsValidator.spec.ts

```diff
@@ -0,0 +1,135 @@
+import _ from 'lodash';
+import SettingsValidator from '../settingsValidator';
+import * as settings from '@/config/settings';
+import { RecursivePartial } from '@/utils/recursivePartialType';
+
+const cfg = settings.init();
+
+if (!cfg.kubernetes.version) {
+  cfg.kubernetes.version = '1.23.4';
+}
+const currK8sVersion = cfg.kubernetes.version;
+const finalK8sVersion = currK8sVersion.replace(/^v?/, '');
+const subject = new SettingsValidator();
+
+subject.k8sVersions = [finalK8sVersion, '1.0.0'];
+describe(SettingsValidator, () => {
+  describe('validateSettings', () => {
+    it('should do nothing when given existing settings', () => {
+      const [needToUpdate, errors] = subject.validateSettings(cfg, cfg);
+
+      expect(needToUpdate).toBeFalsy();
+      expect(errors).toHaveLength(0);
+    });
+
+    it('should want to apply changes when valid new settings are proposed', () => {
+      const newEnabled = !cfg.kubernetes.enabled;
+      const newVersion = subject.k8sVersions[1];
+      const newEngine = cfg.kubernetes.containerEngine === 'moby' ? 'containerd' : 'moby';
+      const newConfig = _.merge({}, cfg, {
+        kubernetes:
+          {
+            enabled:         newEnabled,
+            version:         newVersion,
+            containerEngine: newEngine
+          }
+      });
+      const [needToUpdate, errors] = subject.validateSettings(cfg, newConfig);
+
+      expect(needToUpdate).toBeTruthy();
+      expect(errors).toHaveLength(0);
+    });
+
+    it('should report errors for unchangeable fields', () => {
+      const desiredEnabled = !cfg.kubernetes.enabled;
+      const desiredEngine = cfg.kubernetes.containerEngine === 'moby' ? 'containerd' : 'moby';
+      const requestedSettings = {
+        kubernetes:
+          {
+            enabled:                    desiredEnabled,
+            checkForExistingKimBuilder: !cfg.kubernetes.checkForExistingKimBuilder,
+            containerEngine:            desiredEngine,
+          }
+      };
+      const [needToUpdate, errors] = subject.validateSettings(cfg, requestedSettings);
+
+      expect(needToUpdate).toBeFalsy();
+      expect(errors).toHaveLength(1);
```

Comment:
You don't need a length check if you have a check on all the items.

Reply:
True. For larger arrays I prefer a length check because it's easier to figure out when something goes wrong, but in this case, sure.